iree/turbine/kernel/wave/gather_to_shared.py

```diff
@@ -0,0 +1,367 @@
+    logger.info(f"expected_number_of_loads={expected_number_of_loads}")
+    elements_per_thread = ceildiv(
+        total_number_of_elements, expected_number_of_loads * total_number_of_threads
+    )
+    logger.info(f"elements_per_thread={elements_per_thread}")
+
+    vector_width = elements_per_thread * bitwidth
+    load_width = get_load_width(supported_load_widths, vector_width)
+    if load_width is None:
+        logger.error(f"No supported load width found for bitwidth={vector_width}")
+        return None
+
+    logger.info(f"load_width={load_width}")
+
+    # Get supported load width for the given bitwidth and if they are not
+    # equal then we need to adjust the number of loads and elements per thread.
+    ratio = vector_width // load_width
+    logger.info(f"ratio={ratio}")
+    expected_number_of_loads *= ratio
+    elements_per_thread //= ratio
+
+    return GatherToSharedConfig(
+        materialized_shape,
+        elements_per_thread,
+        expected_number_of_loads,
+    )
+
+
+def create_writes(
+    read: Read,
+    write: Write,
+    materialized_shape: list[IndexSymbol],
+    elements_per_thread: int,
+    expected_number_of_loads: int,
+    total_number_of_threads: int,
+    thread_id: IndexExpr,
+    symbolic_shape: list[IndexSymbol],
+    wave_subs: dict[IndexSymbol, IndexExpr],
+    element_type: "DataType",
+) -> defaultdict[fx.Node, list[Write]]:
+    """
+    Create new writes for the given read and write.
+    """
+    elements_per_wave = elements_per_thread * total_number_of_threads
+    logger.info(f"elements_per_wave={elements_per_wave}")
+
+    materialized_shape_adjusted = list(materialized_shape)
+    materialized_shape_adjusted[-1] = sympy.ceiling(
+        materialized_shape[-1] / elements_per_thread
+    )
+    logger.info(f"materialized_shape_adjusted={materialized_shape_adjusted}")
+
+    # GatherToLDS writes `elements_per_wave` elements contiguously to LDS, so we
+    # cannot have any padding if it crosses a array row boundary.
+    drop_padding = materialized_shape[-1] % elements_per_wave != 0
+
+    global_index = remove_thread_indexing(read.index)
+    logger.info(f"global_index={global_index}")
+
+    new_writes = defaultdict(list)
+    for i in range(expected_number_of_loads):
+        nd_index = delinearize_index(
+            thread_id + i * total_number_of_threads,
+            materialized_shape_adjusted,
+        )
+        logger.info(f"nd_index={nd_index}")
+        write_index = {}
+        for dim, idx in zip(symbolic_shape, nd_index):
+            last = dim == symbolic_shape[-1]
+
+            idx = idx * elements_per_thread if last else idx
+            size = elements_per_thread if last else 1
+            stride = 1
+            write_index[dim] = IndexSequence(idx, size, stride)
+
+        read_index = combine_index(global_index, write_index)
+
+        write_index = {k: v.subs(wave_subs) for k, v in write_index.items()}
```

Comment:
Seems like a non trivial logic/code here, can we add a little comment/intuiton on what we are trying to do here? :)

Reply:
did some comments pass, PTAL